### PR TITLE
[FIX] web: prevent infinite recursion in doActionButton

### PR DESCRIPTION
**Pull request by dise-odoo**
Before this commit, when the current user goes to an action with
embedded actions and create a custom embedded action to load the main
one with custom filters, for instance, and then sorts the embedded
actions displayed in the top bar to first load his own embedded action
instead of the main one. When the user will try to load the main action
to load his custom one by default, the UI will be blocked because of a
infinite recursion.

This is because the condition "action.embedded_actions_ids &&
!this.currentController?.config.currentEmbeddedActionId" is wrong, since
this.currentController is the same object between recursions ; this can
cause an infinite recursive call in some cases.

To fix this, the condition was changed to use flag recently introduced
to know when it is embedded actions calling `doActionButton` to make
sure we will no longer have any infinite recursions inside that function.

Steps to reproduce the issue:
============================

1. Install project app (with demo data)
2. Go to project app
3. Select `Office design` project to see the tasks linked to that
   project.
4. Click on the button to see the top bar containing the embedded
   actions.
5. Remove the `Open tasks` filter and create a custom embedded action.
6. Sort the embedded actions inside the top bar to have the custom one
   first.
7. Go back to the project dashboard.
8. Select again `Office design` project.

Expected behavior:
-----------------

The custom embedded action should be loaded instead of the main one
without any issues.

Current Behavior:
-----------------

The UI completely freezes because of an infinite recursions in the JS
code.

task-4283994

### CODE REVIEW

File: addons/web/static/src/webclient/actions/action_service.js [hunk 1]
```diff
@@ -1446,7 +1446,7 @@ export function makeActionManager(env, router = _router) {
             }
             throw new InvalidButtonParamsError("Missing type for doActionButton request");
         }
-        if (action.embedded_action_ids && !this.currentController?.config.currentEmbeddedActionId) {
+        if (params.context?.current_embedded_action_id === undefined) {
```
Comment (deneuvillem) [inline addons/web/static/src/webclient/actions/action_service.js:1449]: This condition is true in two cases :
1) params.context is undefined or does not exist
2) params.context.current_embedded_action_id is explicitly undefined

Is this intended to execute the block if 1) is true ?

**Review by deneuvillem — COMMENTED** [review]
Thanks for your fix :muscle: I just left one question for you :smile:

File: addons/web/static/src/webclient/actions/action_service.js (hunk 1 above)
Comment (dise-odoo) [reply addons/web/static/src/webclient/actions/action_service.js:1449]: No, it is not intended. Thanks for pointing it out :)

**Comment by xavierbol** [comment]
We could also add a test to make sure we will no longer have that annoying issue in the future. 🙂

**Comment by dise-odoo** [comment]
> We could also add a test to make sure we will no longer have that annoying issue in the future. 🙂

@xavierbol I am not sure this is testable, how can I make a test that detects an infinite recursion? :D

**Review by deneuvillem — COMMENTED** [review]
LGTM :100: Thanks for your work! :1st_place_medal: 
Adding a test could be nice, but as you said, it seems to be quite difficult to test infinite recursion :thinking:

**Review by xavierbol — APPROVED** [review]
LGTM, thanks for your work! 🙂 

For the test, maybe we could check the number of times doActionButton is called and if it is more than 2 for instance it means we have an issue and surely the infinite loop? 🤷🏼‍♂️

**Comment by xavierbol** [comment]
Coucou @odoo/rd-framework-js 

Can we have your opinion? 

We don't really know if it is possible to add a test for that issue. Do you have an idea to correctly test it? 🙂 

Thanks in advance!

Have a nice day! 🙂

**Comment by pparidans** [comment]
What about creating a test that recreates exactly the steps to reproduce this issue (cf. "custom embedded action and setting it first in the top bar") and see if it steadily fails ?

Seems naive but would be an interesting starting point. Or maybe is it non-determinist?

**Comment by aab-odoo** [comment]
@dise-odoo What about something like this? I think it's less obscur... Also, I'll take some time to dig into the embedded action logic and maybe try to rework it a bit in master. I think we could do a bit better API-wise than what we have now.
```diff
diff --git a/addons/web/static/src/search/control_panel/control_panel.js b/addons/web/static/src/search/control_panel/control_panel.js
index dca7a4968d6d..ee869686d89d 100644
--- a/addons/web/static/src/search/control_panel/control_panel.js
+++ b/addons/web/static/src/search/control_panel/control_panel.js
@@ -567,7 +567,7 @@ export class ControlPanel extends Component {
             parent_action_embedded_actions: this.state.embeddedInfos.embeddedActions,
             parent_action_id: action.parent_action_id[0] || action.parent_action_id,
         };
-        this.actionService.doActionButton({
+        this.actionService.executeEmbeddedAction({
             type: action.python_method ? "object" : "action",
             resId: this.env.searchModel?.globalContext.active_id,
             name: action.python_method || action.action_id[0] || action.action_id,
diff --git a/addons/web/static/src/webclient/actions/action_service.js b/addons/web/static/src/webclient/actions/action_service.js
index 25b8a6074185..2406cf41d0b7 100644
--- a/addons/web/static/src/webclient/actions/action_service.js
+++ b/addons/web/static/src/webclient/actions/action_service.js
@@ -1396,7 +1396,7 @@ export function makeActionManager(env, router = _router) {
      * @param {DoActionButtonParams} params
      * @returns {Promise<void>}
      */
-    async function doActionButton(params) {
+    async function doActionButton(params, canExecuteEmbeddedAction = true) {
         // determine the action to execute according to the params
         let action;
         const context = makeContext([params.context, params.buttonContext]);
@@ -1446,7 +1446,7 @@ export function makeActionManager(env, router = _router) {
             }
             throw new InvalidButtonParamsError("Missing type for doActionButton request");
         }
-        if (action.embedded_action_ids && !this.currentController?.config.currentEmbeddedActionId) {
+        if (action.embedded_action_ids && canExecuteEmbeddedAction) {
             const embeddedActionsOrder = JSON.parse(
                 browser.localStorage.getItem(
                     `orderEmbedded${action.id}+${params.resId || ""}+${user.userId}`
@@ -1477,7 +1477,7 @@ export function makeActionManager(env, router = _router) {
                     parent_action_embedded_actions: embeddedActions,
                     parent_action_id: action.id,
                 };
-                await this.doActionButton({
+                return this.executeEmbeddedAction({
                     name:
                         embeddedAction.python_method ||
                         embeddedAction.action_id[0] ||
@@ -1488,7 +1488,6 @@ export function makeActionManager(env, router = _router) {
                     resModel: embeddedAction.parent_res_model,
                     viewType: embeddedAction.default_view_mode,
                 });
-                return;
             }
         }
         // filter out context keys that are specific to the current action, because:
@@ -1523,6 +1522,10 @@ export function makeActionManager(env, router = _router) {
         }
     }
 
+    function executeEmbeddedAction(params) {
+        return this.doActionButton(params, false);
+    }
+
     /**
      * Switches to the given view type in action of the last controller of the
      * stack. This action must be of type 'ir.actions.act_window'.
@@ -1700,6 +1703,7 @@ export function makeActionManager(env, router = _router) {
     return {
         doAction,
         doActionButton,
+        executeEmbeddedAction,
         switchView,
         restore,
         loadState,

```

**Comment by aab-odoo** [comment]
This can indeed quite easiliy be tested

**Comment by xavierbol** [comment]
I added the test to reproduce the issue. 🙂 

I think now we are good to go, right? 🙂

File: addons/web/static/tests/webclient/actions/embedded_action.test.js [hunk 2]
```diff
@@ -481,3 +501,26 @@ test("execute a regular action from an embedded action", async () => {
...
+    });
+    expect(".o_list_view").toHaveCount(1);
+    await contains(".o_control_panel_navigation > button > i.fa-sliders").click();
+    expect(".o_embedded_actions > button:first-child").toHaveClass("active", {
+        message: "First embedded action in order should have the 'active' class",
+    });
+    expect(".o_embedded_actions > button:first-child > span").toHaveText(
+        "Custom Embedded Action 4",
+        {
+            message: "First embedded action in order should be 'Embedded Action 2'",
```
Comment (aab-odoo) [inline addons/web/static/tests/webclient/actions/embedded_action.test.js:520]:
```suggestion
            message: "First embedded action in order should be 'Embedded Action 4'",
```

**Review by aab-odoo — APPROVED** [review]
robodoo delegate+

Thanks @xavierbol

File: addons/web/static/tests/webclient/actions/embedded_action.test.js (hunk 2 above)
Comment (xavierbol) [reply addons/web/static/tests/webclient/actions/embedded_action.test.js:520]: Oups, indeed! 

Thanks!

**Comment by xavierbol** [comment]
robodoo r+

**Pull request closed without merging**